src/schedule/zephyr_dp_schedule.c

```diff
@@ -0,0 +1,376 @@
+		curr_task = container_of(tlist, struct task, list);
+		pdata = curr_task->priv_data;
+
+		if (pdata->ticks_to_trigger == 0) {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED) {
+				/* set new trigger time, start the thread */
+				pdata->ticks_to_trigger = pdata->ticks_period;
+				curr_task->state = SOF_TASK_STATE_RUNNING;
+				k_sem_give(&pdata->sem);
+			}
+		} else {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED ||
+			    curr_task->state == SOF_TASK_STATE_RUNNING)
+				/* decrease num of ticks to re-schedule */
+				pdata->ticks_to_trigger--;
+		}
+	}
+	zephyr_dp_unlock(&lock_flags);
+}
+
+static int scheduler_dp_task_cancel(void *data, struct task *task)
+{
+	(void)(data);
+	uint32_t lock_flags;
+	struct dp_task_pdata *pdata = task->priv_data;
+
+	/* this is asyn cancel - mark the task to terminate */
+	zephyr_dp_lock(&lock_flags);
+	task->state = SOF_TASK_STATE_CANCEL;
+	list_item_del(&task->list);
+	zephyr_dp_unlock(&lock_flags);
+
+	/* release the thread to let it terminate itself */
+	k_sem_give(&pdata->sem);
+
+	return 0;
+}
+
+static int scheduler_dp_task_free(void *data, struct task *task)
+{
+	uint32_t lock_flags;
+	struct dp_schedule_data *sch = data;
+	struct dp_task_pdata *pdata = task->priv_data;
+
+	/* abort the execution of the thread */
+	k_thread_abort(pdata->thread_id);
+
+	zephyr_dp_lock(&lock_flags);
+	list_item_del(&task->list);
+	task->priv_data = NULL;
+	task->state = SOF_TASK_STATE_FREE;
+	zephyr_dp_unlock(&lock_flags);
+
+	/* free all allocated resources */
+	rfree(pdata->p_stack);
+	rfree(pdata->thread_id);
+	rfree(pdata);
+
+	return 0;
+}
+
+/* Thread function called in component context, on target core */
+static void dp_thread_fn(void *p1, void *p2, void *p3)
+{
+	struct task *task = p1;
+	(void) p2;
+	(void) p3;
+	struct dp_task_pdata *task_pdata = task->priv_data;
+	uint32_t lock_flags;
+	enum task_state state;
+
+	do {
+		/*
+		 * the thread is started immediately after creation, it will stop on semaphore
+		 * Semaphore will be released once the task is ready to process
+		 */
+		k_sem_take(&task_pdata->sem, K_FOREVER);
+
+		if (task->state == SOF_TASK_STATE_RUNNING) {
+			state = task_run(task);
+
+			/*
+			 * handle task state - the task may be canceled by external call
+			 * or request termination itself
+			 */
+			zephyr_dp_lock(&lock_flags);
+
+			/*
+			 * check if task is still running, may have be canceled by external call
+			 * if not, set the state returned by run procedure
+			 */
+			if (task->state == SOF_TASK_STATE_RUNNING)
+				task->state = state;
+
+			switch (task->state) {
+			case SOF_TASK_STATE_RESCHEDULE:
+				/* mark to reschedule, schedule time is already calculated */
+				task->state = SOF_TASK_STATE_QUEUED;
+				break;
+
+			case SOF_TASK_STATE_CANCEL:
+				/* mark the task as cancelled */
+				scheduler_dp_task_cancel(&dp_sch, task);
+				break;
+
+			case SOF_TASK_STATE_COMPLETED:
+				break;
+
+			default:
+				/* illegal state */
+				assert(false);
```

Comment:
I can see why you wanted to raise the assert discussion. ;) But agreed, this is definitely good example of assert that would be good to leave in product builds. OTOH, one could argue along the lines of https://www.kernel.org/doc/Documentation/process/coding-style.rst "Do not crash the kernel" , that even with state corruption, system should not be crashes if not absolutely necessary. We do have sof_panic() and k_panic() available for those cases where system should absolutely panic and that's something you could use here as well and it'd always be in all builds.

Reply:
> this is definitely good example of assert that would be good to leave in product builds.

Let's stick to most common industry terminology and call that a `panic()` or `BUG_ON()`,  `error()`, etc. By definition `assert()`s are turned off to save cycles or memory.

Reply:
FWIW, the relevant Zephyr APIs are `k_panic()` (intended for fatal system-wide errors) and `k_oops()` (thread-terminating errors).  In fact both do the same thing, and arrive in `k_sys_fatal_error_handler()` with a "reason" code of `K_ERR_KERNEL_{OOPS,PANIC}`.   Despite the k_ prefix, that function is actually application-defined and can do whatever you want.  The default is a weak function in kernel/fatal.c that just logs the panic (but doesn't dump the stack, we should add that) and halts the system.